lib/multi-writable.js

```diff
@@ -59,43 +120,260 @@ class MultiWriteable extends stream.Writable {
+
+    if (indexOfCreate > -1 && (indexOfInsert === -1 || indexOfCreate < indexOfInsert)) {
+      this.state = STATES.CREATE;
+      this.runReadSchema();
+    } else if (indexOfInsert > -1 && (indexOfCreate === -1 || indexOfInsert < indexOfCreate)) {
+      this.state = STATES.INSERT;
+      this.runInsert();
+    } else {
+      this.state = null;
+      this.next();
+    }
   }
 
-  flushQueueToAvailableWorkers(next) {
-    const nextWorkers = this.readyWorkers.splice(0, this.queue.length);
+
+  // Handle Insert State
+
+  setTable(tableName) {
+    if (this.tables[tableName]) {
+      this.currentTable = tableName;
+    } else {
+      throw new Error(`The schema for table '${tableName}' has not been set yet. Your dump needs a CREATE TABLE query.`);
+    }
+  }
+
+  runInsert() {
+    const indexOfInsert = this.buffer.indexOf(INSERT_INTO_BUFF, this.bufferPosition);
+    const indexOfValues = this.buffer.indexOf(VALUES_BUFF, indexOfInsert);
+
+    if (indexOfValues === -1) {
+      this.next();
+    } else {
+      const tablePos = indexOfInsert + INSERT_INTO_BUFF.length;
+      const tableName = cleanVarName(this.buffer.slice(tablePos, indexOfValues).toString());
+      this.setTable(tableName);
+
+      this.bufferPosition = indexOfValues + VALUES_BUFF.length;
+      this.continueInsert();
+    }
+  }
+
+  continueInsert() {
     const writes = [];
-    let queueItem;
     let row;
-
-    for (let i = 0; i < nextWorkers.length; i += 1) {
-      queueItem = this.queue.shift();
-      row = this.processQueueItem(queueItem);
-      writes.push(this.writeToNextAvailableWorker(nextWorkers[i], row));
+    let worker;
+    let queueItem = this.getNextCommandParenSet();
+
+    while (
+      queueItem
+      && queueItem !== COMMAND_EXEC_BUFF
+      && this.readyWorkers.length
+    ) {
+      worker = this.readyWorkers.pop();
+      row = `[${queueItem}]`;
+      writes.push(this.writeToNextAvailableWorker(worker, row));
+
+      queueItem = this.getNextCommandParenSet();
     }
 
     Promise
       .all(writes)
       .then(() => {
-        if (this.queue.length) {
-          return this.flushQueueToAvailableWorkers(next);
+        if (!queueItem && this.bufferPosition >= this.buffer.length - 1) {
+          return this.next();
+        }
+        if (queueItem === COMMAND_EXEC_BUFF) {
+          this.cleanBuffer();
+          return this.determineState();
         }
 
-        return next();
+        return this.continueInsert();
       })
       .catch((err) => {
         console.error(err);
       });
   }
 
+
+  // Handle Create Table State
+
+  runReadSchema() {
+    const indexOfCreate = this.buffer.indexOf(CREATE_TABLE_BUFF);
+    const indexOfNextParen = this.buffer.indexOf(BRACKET_OPEN_BUFF, indexOfCreate);
+    const startPos = indexOfCreate + CREATE_TABLE_BUFF.length;
+
+    // set buffer head to the end of values
+    this.bufferPosition = startPos;
+
```

Comment:
This method needs better handling of unfulfilled command in buffer to call this.next() if missing content.